internal/charmstore/migrations.go

```diff
@@ -222,3 +227,93 @@ func populateWriteACL(db StoreDatabase) error {
 	logger.Infof("%d base entities updated", counter)
 	return nil
 }
+
+func populatePromulgatedEntities(db StoreDatabase) error {
+	baseEntities := db.BaseEntities()
+	entities := db.Entities()
+
+	// 1. Update all base entities to have a promulgated value of the correct form.
+	_, err := baseEntities.UpdateAll(
+		bson.D{{"promulgated", bson.D{{"$ne", 1}}}},
+		bson.D{{"$set", bson.D{{"promulgated", -1}}}},
+	)
+	if err != nil {
+		return errgo.Notef(err, "cannot set promulgated to initial false value")
+	}
+
+	// 2. Update entities with users with their promulgated URL.
+	//
+	// Note that we order the query results by revision, so when
+	// we've traversed all the entities, owner[name] will hold the
+	// owner of the latest revision of each promulgated charm with
+	// the given name.
+	owners := make(map[string]string)
+	iter := entities.Find(bson.D{{"user", ""}}).Select(bson.D{
+		{"_id", 1},
+		{"baseurl", 1},
+		{"blobhash", 1},
+		{"extrainfo.bzr-owner", 1},
+	}).Sort("name", "revision").Iter()
+	var entity mongodoc.Entity
+	for iter.Next(&entity) {
+		var user string
+		if err := json.Unmarshal([]byte(entity.ExtraInfo["bzr-owner"]), &user); err != nil {
```

Comment:
If that's the case we should fix that first, lets make the data good rather than the code complicated.